scalar/float.go

```diff
@@ -7,160 +7,52 @@ import (
-	return s.Value
-}
-
-func (s *Float32) Set(val any) error {
-	if val == nil {
-		s.Valid = false
-		return nil
-	}
-
-	if sc, ok := val.(Scalar); ok {
-		if !sc.IsValid() {
-			s.Valid = false
-			return nil
-		}
-		return s.Set(sc.Get())
-	}
-
-	switch value := val.(type) {
-	case int8:
-		s.Value = float32(value)
-	case int16:
-		s.Value = float32(value)
-	case int32:
-		s.Value = float32(value)
-	case int64:
-		if value > math.MaxInt32 {
-			return &ValidationError{Type: arrow.PrimitiveTypes.Float32, Msg: "int64 bigger than MaxInt8", Value: value}
-		}
-		s.Value = float32(value)
-	case uint8:
-		s.Value = float32(value)
-	case uint16:
-		s.Value = float32(value)
-	case uint32:
-		if value > math.MaxInt32 {
-			return &ValidationError{Type: arrow.PrimitiveTypes.Int8, Msg: "uint32 bigger than MaxInt8", Value: value}
-		}
-		s.Value = float32(value)
-	case uint64:
-		if value > math.MaxInt32 {
-			return &ValidationError{Type: arrow.PrimitiveTypes.Int8, Msg: "uint64 bigger than MaxInt8", Value: value}
-		}
-		s.Value = float32(value)
-	case float32:
-		s.Value = value
-	case float64:
-		if value > math.MaxInt32 {
-			return &ValidationError{Type: arrow.PrimitiveTypes.Int8, Msg: "float64 bigger than MaxInt8", Value: value}
-		}
-		s.Value = float32(value)
-	case string:
-		num, err := strconv.ParseFloat(value, 32)
-		if err != nil {
-			return &ValidationError{Type: arrow.PrimitiveTypes.Int8, Msg: "invalid string", Value: value}
-		}
-		s.Value = float32(num)
-	case *int8:
-		return s.Set(*value)
-	case *int16:
-		return s.Set(*value)
-	case *int32:
-		return s.Set(*value)
-	case *int64:
-		return s.Set(*value)
-	case *uint8:
-		return s.Set(*value)
-	case *uint16:
-		return s.Set(*value)
-	case *uint32:
-		return s.Set(*value)
-	case *uint64:
-		return s.Set(*value)
-	case *float32:
-		return s.Set(*value)
-	case *float64:
-		return s.Set(*value)
+func (s *Float) DataType() arrow.DataType {
+	switch s.getBitWidth() {
+	case 64:
+		return arrow.PrimitiveTypes.Float64
+	case 32:
+		return arrow.PrimitiveTypes.Float32
+	case 16:
+		return arrow.FixedWidthTypes.Float16
 	default:
-		if originalSrc, ok := underlyingNumberType(value); ok {
-			return s.Set(originalSrc)
-		}
-		return &ValidationError{Type: arrow.PrimitiveTypes.Int8, Msg: noConversion, Value: value}
+		panic("invalid bit width")
 	}
-	s.Valid = true
-	return nil
 }
 
-type Float64 struct {
-	Valid bool
-	Value float64
-}
-
-func (s *Float64) IsValid() bool {
-	return s.Valid
-}
-
-func (*Float64) DataType() arrow.DataType {
-	return arrow.PrimitiveTypes.Float64
-}
-
-func (s *Float64) Get() any {
+func (s *Float) Get() any {
 	return s.Value
 }
 
-func (s *Float64) Equal(rhs Scalar) bool {
+func (s *Float) Equal(rhs Scalar) bool {
 	if rhs == nil {
 		return false
 	}
-	r, ok := rhs.(*Float64)
+	r, ok := rhs.(*Float)
 	if !ok {
 		return false
 	}
-	return s.Valid == r.Valid && s.Value == r.Value
+	return s.getBitWidth() == r.getBitWidth() && s.Valid == r.Valid && s.Value == r.Value
```

Comment:
are we sure we want the comparison to depend on the bit width?
what about float16(0) vs float32(0) vs float64(0)?

Reply:
They are separate types in the arrow world so I'm guessing that's OK :)